### PR TITLE
Update upgrade database worker with the correct dependencies

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following upgrades to the new predicate watchers for the upgrade 
service. This simplifies the watchers for upgrades dramatically. We're
no longer running bespoke watchers for upgrades. The watcher there
was already out of date with the new learnings from the London sprint.

This is more foundational work on getting the right information into the
worker, which then can assess how to move forward.

--------

The code is rather mechanical, we're just pushing dependencies to
where they need to go. I did make the upgradeDBWorker a catacomb
instead of tomb, as this will need to have associated watchers and that
requires lifecycle management.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
$ juju add-model default
$ juju deploy ubuntu
$ juju upgrade-controller --build-agent
```

## Links

**Jira card:** JUJU-4754

### CODE REVIEW

File: domain/upgrade/service/service.go [hunk 1]
```diff
@@ -116,13 +118,17 @@ func (s *Service) ActiveUpgrade(ctx context.Context) (domainupgrade.UUID, error)
...
+// WatchForUpgradeStarted creates a watcher which notifies when all controller
+// nodes have been registered, meaning the upgrade is ready to start.
+func (s *Service) WatchForUpgradeStarted(ctx context.Context, upgradeUUID domainupgrade.UUID) (watcher.NotifyWatcher, error) {
 	if err := upgradeUUID.Validate(); err != nil {
 		return nil, errors.Trace(err)
 	}
-	return NewUpgradeReadyWatcher(ctx, s.st, s.watcherFactory, upgradeUUID)
+
+	mask := changestream.Create | changestream.Update
+	return s.watcherFactory.NewValuePredicateWatcher("upgrade_info_controller_node", upgradeUUID.String(), mask, func(ctx context.Context, db coredatabase.TxnRunner, changes []changestream.ChangeEvent) (bool, error) {
```
Comment (manadart) [inline domain/upgrade/service/service.go:129]: Could declare the predicate func separately or assign it to a variable to shorten this line.

File: domain/upgrade/service/service.go [hunk 2]
```diff
@@ -116,13 +118,17 @@ func (s *Service) ActiveUpgrade(ctx context.Context) (domainupgrade.UUID, error)
 	return activeUpgrade, domain.CoerceError(err)
 }
 
-// WatchForUpgradeReady creates a watcher which notifies when all controller
-// nodes have been registered, meaning the upgrade is ready to start
-func (s *Service) WatchForUpgradeReady(ctx context.Context, upgradeUUID domainupgrade.UUID) (watcher.NotifyWatcher, error) {
+// WatchForUpgradeStarted creates a watcher which notifies when all controller
```
Comment (manadart) [inline domain/upgrade/service/service.go:121]: `WatchForUpgradeReady` is a better name.

File: worker/upgradedatabase/manifold.go [hunk 3]
```diff
@@ -45,11 +59,13 @@ func Manifold(cfg ManifoldConfig) dependency.Manifold {
...
 			cfg.AgentName,
 			cfg.UpgradeDBGateName,
+			cfg.ServiceFactoryName,
+			cfg.DBAccessorName,
 		},
 		Start: func(context dependency.Context) (worker.Worker, error) {
 			// Get the completed lock.
-			var upgradeDatabaseLock gate.Lock
-			if err := context.Get(cfg.UpgradeDBGateName, &upgradeDatabaseLock); err != nil {
+			var upgradeCompleteLock gate.Lock
```
Comment (manadart) [inline worker/upgradedatabase/manifold.go:67]: I'd go a little more specific with `dbUpgradeCompleteLock`.

File: worker/upgradedatabase/worker.go [hunk 4]
```diff
@@ -65,24 +89,47 @@ func (c Config) Validate() error {
...
+		Work: w.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
 	}
 
-	w.tomb.Go(w.loop)
 	return w, nil
 }
 
+// Kill implements worker.Worker.Kill.
```
Comment (manadart) [inline worker/upgradedatabase/worker.go:121]: I still think we should do with workers what we did with the watcher base and embed a catacomb/tomb worker that hoists these. Different patch though.

**Review by manadart — APPROVED** [review]
Good to go; consider the naming changes.

File: worker/upgradedatabase/worker.go (hunk 4 above)
Comment (SimonRichardson) [reply worker/upgradedatabase/worker.go:121]: I have started doing this in another patch elsewhere. The problematic thing to consider is the loop/run part. It requires you to know upfront where the loop will be.

**Pull request merged**